Feature/js refactor 1

achen2401 opened the pull request:
https://jira.movember.com/browse/TN-483
remove dependencies among core objects (e.g. tnthAjax, assembleContent, etc.)
move profile object out from mainJS
use Vue when necessary to render data
refactor profile templates to remove duplicate code
add gulpfile for development tasks, e.g. compiling less, minify code, etc.
bug fixes in initial queries
also containing bug fixes for:  TN-1025, TN-988, TN-1035

I did add TODO notes for future work, this is initial foray but hopefully a good start
# WILL BE FOR NEXT SPRINT, DON'T MERGE YET, WORKING ON FIXES BASED ON COMMENTS

ivan-c commented on portal/static/js/main.js:
```diff
@@ -5970,7 +2452,7 @@ var tnthDates = {
      * changeFormat("2016-04-29",true,true) returns "29/04/2016"
      ***/
-    "changeFormat": function(currentDate,reverse,shorten) {
+    "changeFormat": function(currentDate, reverse, shorten) {
```
This might be simpler if you break the parsing and the formatting into 2 different functions

ivan-c commented on portal/static/js/main.js:
```diff
@@ -6196,13 +2678,13 @@ var tnthDates = {
-            returnDate = ('0' + todayDate.getDate()).slice(-2)+" de "+monthNames[todayDate.getMonth()]+" de "+todayDate.getFullYear();
-        } else if(userSetLang !== undefined && userSetLang == "en_AU") {
-            returnDate = ('0' + todayDate.getDate()).slice(-2)+" "+monthNames[todayDate.getMonth()]+" "+todayDate.getFullYear();
+        if (userSetLang !== undefined && userSetLang == "es_MX") {
```
Can this be replaced with i18next code?

ivan-c commented on portal/static/js/main.js:
```diff
@@ -5219,699 +1707,695 @@ var tnthAjax = {
     }
 };
+
 var tnthDates = {
```
Could you move this to a separate dependency, or use a library like momentJS?

ivan-c commented on portal/static/js/procedures.js:
```diff
@@ -1,289 +1,393 @@
-                            };
+(function() {
+    //TODO this JS code file needs a bit of clean-up
+    var CANCER_TREATMENT_CODE = "118877007", NONE_TREATMENT_CODE = "999";
```
Move to code system module?

ivan-c commented on portal/static/js/procedures.js:
```diff
@@ -1,289 +1,393 @@
+                type: "GET",
+                url: "/patients/treatment-options",
+                cache: true
+            }).done(function(data) {
```
Cache in localStorage?

ivan-c commented on portal/static/js/procedures.js:
```diff
@@ -1,289 +1,393 @@
+                cache: true
+            }).done(function(data) {
+                if (data.treatment_options) {
+                    if (data.treatment_options) {
```
Duplicate line?

ivan-c commented on portal/static/js/procedures.js:
```diff
@@ -1,289 +1,393 @@
+            }
+            var dTest = procDateReg.test(d), mTest = (String(m) !== ""), yTest = procYearReg.test(y);
+            var errorText = i18next.t("The procedure date must be valid and in required format.");
+            var dgField = $("#procDateGroup"), deField = $("#procDateErrorContainer"), errorColor = "#a94442", validColor = "#777";
```
Why not add a class instead of a style?

ivan-c commented on portal/static/js/procedures.js:
```diff
@@ -1,289 +1,393 @@
-            if (date.getFullYear() == parseInt(y) && ((date.getMonth() + 1) == parseInt(m)) && date.getDate() == parseInt(d)) {
-                if (date.setHours(0,0,0,0) > today.setHours(0,0,0,0)) {
-                    deField.text(i18next.t("The procedure date must be in the past.")).css("color", errorColor);
+        function checkDate() {
```
Is there any better way to generically validate dates/times?

ivan-c commented on portal/templates/admin_base.html:
```diff
@@ -0,0 +1,8 @@
+{% block additional_scripts %}
+<script src="https://cdnjs.cloudflare.com/ajax/libs/jasny-bootstrap/3.1.3/js/jasny-bootstrap.min.js"></script>
+<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-table/1.11.1/extensions/export/bootstrap-table-export.min.js"></script>
+<script src="//rawgit.com/hhurz/tableExport.jquery.plugin/master/tableExport.js" async></script>
```
We'll have to use [something else](https://alternativeto.net/software/rawgit/) (as suggested by the review). Could you pin this to a [specific release](https://github.com/hhurz/tableExport.jquery.plugin/releases)? master will change over time

ivan-c commented on portal/templates/sessionReport.html:
```diff
@@ -2,47 +2,73 @@
-		{% endif %}
-		&nbsp;&nbsp;<a href="{{PORTAL}}/home" data-toggle="tooltip" data-placement="top" title="{{ _('Back to Truenth Home') }}" class="btn"><span class="glyphicon glyphicon-home" aria-hidden="true"></span></a>
-	</span>
+			&nbsp;&nbsp;<a href="{{PORTAL}}/home" data-toggle="tooltip" data-placement="top" title="{{ _('Back to Truenth Home') }}" class="btn"><span class="glyphicon glyphicon-home" aria-hidden="true"></span></a>
```
Would `url_for()` here work?

achen2401 commented:
@ivan-c Ivan, I think I addressed most of your comments - let me know if you think more changes are needed.  If so, I will make notes and probably just do them in a different PR - want to take a break from this a little :-P

ivan-c commented on portal/static/js/procedures.js:
```diff
@@ -1,289 +1,394 @@
+                            creator = i18next.t("staff member") + ", <span class='creator'>" + (creatorDisplay ? creatorDisplay : creator) + "</span>, ";
+                        }
+                        var dtEdited = val.resource.meta.lastUpdated, dateEdited = new Date(dtEdited);
+                        var creationText = i18next.t("(date entered by %actor on %date)").replace("%actor", creator).replace("%date", dateEdited.toLocaleDateString("en-GB", {
```
Can you make the placeholders consistent with the rest (eg `{foo}`)?

ivan-c commented on portal/static/js/procedures.js:
```diff
@@ -1,289 +1,394 @@
+        function isLeapYear(year) {
+            return ((year % 4 == 0) && (year % 100 != 0)) || (year % 400 == 0);
+        }
+        function checkDate() {
```
Could this be moved to a more general place, like `utility.js`?

ivan-c commented on portal/static/js/procedures.js:
```diff
@@ -1,289 +1,394 @@
+            $(this).parents("tr").fadeOut("slow", function() {
+                $(this).remove(); // Remove from list
+                if ($("#eventListtnthproc tr").length === 0) { //If there's no events left, add status msg back in
+                    $("body").find("#userProcedures").html("<p id='noEvents' style='margin: 0.5em 0 0 1em'><em>You haven't entered any treatments yet.</em></p>").animate({
```
Aha! The one missing translation tag :)

ivan-c commented on portal/static/js/main.js:
```diff
@@ -3966,458 +253,228 @@ OrgTool.prototype.populateOrgsList = function(items) {
-      return orgList;
+    if (!orgs || (orgs.length === 0)) {
+        return orgList;
     } else {
```
I don't think you need this `else`, since the `if` block exits early, this code will never be reached

ivan-c commented on portal/static/js/main.js:
```diff
@@ -4431,441 +488,418 @@ var tnthAjax = {
+            }
+        }
+    },
+    "getCurrentUser": function(callback) {
```
Move to `utility.js`?

ivan-c commented on portal/static/js/main.js:
```diff
@@ -4431,441 +488,418 @@ var tnthAjax = {
         self.sendRequest(__url, "GET", userId, {sync: true}, function(data) {
             if (data) {
                 if (!data.error) {
                     if (data.consent_agreements) {
```
These multiple `if`s might be better to bundle into a single `if` statement

ivan-c commented on portal/static/js/main.js:
```diff
@@ -4875,106 +909,95 @@ var tnthAjax = {
+        var self = this, OT = this.getOrgTool(), userId = $("#fillOrgs").attr("userId") || $("#userOrgs").attr("userId");
         var configVar = $("#profile_CONSENT_WITH_TOP_LEVEL_ORG").val();
-        if (!hasValue(configVar)) {
+        if (!configVar) {
```
You might want to invert this, checking localeStorage should be quicker than checking the DOM

ivan-c commented on portal/static/js/main.js:
```diff
@@ -4875,106 +909,95 @@ var tnthAjax = {
+                        }
+                    }
                 } else {
                     if (cto) {
```
Might be more concise to only set the variable you're looping over inside the if/else, and move the loop code outside

ivan-c commented on portal/static/js/main.js:
```diff
@@ -5257,666 +1189,585 @@ var tnthAjax = {
+        var code = "", display = "";
+        switch (toCall) {
+        case "biopsy":
+            code = "111";
```
Should these be part of `SYSTEM_IDENTIFIER_ENUM`?

ivan-c commented on portal/static/js/initialQueries.js:
```diff
@@ -1,1119 +1,1062 @@
-                            $(this).attr("type", "radio");
-                            if (checkedOrgs[$(this).val()]) $(this).prop("checked", true);
+                    if (!hasValue($("#iqPatientEditable").val())) { //for patient, clinic is drawn in orgs state selector template
+                        tnthAjax.getOrgs($("#iq_userId").val(), true, function() {
```
Can you get the current user from localStorage?

ivan-c commented on portal/static/js/profile.js:
```diff
@@ -0,0 +1,2784 @@
+                }
+            }, 10);
+        },
+        data: {
```
Maybe it would make sense to move this to a separate file?

ivan-c commented on portal/static/js/profile.js:
```diff
@@ -0,0 +1,2784 @@
+                var isEditableByPatient = this.settings.hasOwnProperty("CONSENT_EDIT_PERMISSIBLE_ROLES") && this.settings.CONSENT_EDIT_PERMISSIBLE_ROLES.indexOf("patient") !== -1;
+                return this.isAdmin() || (isStaff && isEditableByStaff) || (isPatient && isEditableByPatient);
+            },
+            isTestPatient: function() {
```
this might be better checked directly, could be misconstrued with `test` role...

ivan-c commented on portal/static/js/profile.js:
```diff
@@ -0,0 +1,2784 @@
+                                $("#passwordResetMessage").text(i18next.t("Unable to send email."));
+                            }
+                        });
+                 
```
Missed a translation tag :)

ivan-c commented:
@achen2401 Is it OK if I merge what you have, and you address comments in subsequent PRs?

achen2401 commented:
@ivan-c  ivan, per our chat, I will ask Sie-ce to test using my instance first.  These will most likely be ready to go after next Monday.

achen2401 commented:
merging this per chat with Justin and Ivan
